### ISSUE TITLE
Stream video example shows switched RGB colors (red <-> blue)

### ISSUE THREAD

**Issue by nunoguedelha** [robotology/yarp.js#41]
When running the stream video example and streaming the Framegrabber "bouncing" ball or the scrolling "line", we can clearly see that the RGB colors are swapped from RGB to BGR.

> By using the bouncing ball mode (option `--mode ball`) and change the code section https://github.com/robotology/yarp/blob/82bcafc792ba9b32aad771f7f5d6f3c52ba5fbbe/src/devices/fakeFrameGrabber/FakeFrameGrabber.cpp#L434-L442 for drawing 3 RGB circles instead of two
> ```c++
>     case VOCAB_BALL:
>         {
>             if (have_bg) {
>                 image.copy(background);
>             } else {
>                 image.zero();
>             }
>             addCircle(image,PixelRgb{255,0,0},bx,by,22);
>             addCircle(image,PixelRgb{0,255,0},bx,by,15);
>             addCircle(image,PixelRgb{0,0,255},bx,by,8);
> ```
> 
> we can check the color order used on the receiver side to be **BGR**:
> 1. Run the `fakeFrameGrabber`
>   ```
>   yarpdev --file fakeFrameGrabber_basic.ini --mode ball --name /icubSim/camLeftEye
>   ```
>   `fakeFrameGrabber_basic.ini`:
>   ```
>   device fakeFrameGrabber
>   width 640
>   height 480
>   period 10
>   syncro 1
>   ```
> 2. Connect the the output port to a `yarpview` device.
> 3. Run the Open-MCT visualizer.
> 
> <img width="1125" alt="image" src="https://user-images.githubusercontent.com/6848872/160327942-45aae3fc-11a0-49c4-abc8-3d6f565f0b36.png">
> 

_Originally posted by @nunoguedelha in https://github.com/ami-iit/yarp-openmct/issues/101#issuecomment-1080179870_

**Comment by nunoguedelha** [comment]
## Understand how the Color Space is configured on the Canvas API 
Part of the [canvas API](https://developer.mozilla.org/en-US/docs/Web/API/Canvas_API) is addressed in https://github.com/robotology/yarp.js/issues/39#issuecomment-1085737849. The `CanvasRenderingContext2D` interface, part of the **Canvas API**, provides the 2D rendering context for the drawing surface of a [<canvas>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/canvas) element. It is used for drawing shapes, text, images, and other objects.

We can check the original color space of this 2D rendering context by drawing a rectangle (`[fillRect](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/fillRect)`) filling the whole canvas, of a single color, red (`[fillStyle](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/fillStyle) = "#FF0000"`), and then checking the image data ([`getImageData`](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/getImageData) returns an [`ImageData`](https://developer.mozilla.org/en-US/docs/Web/API/ImageData) object, a one-dimensional array containing the data in the RGBA order, with integer values between 0 and 255).
```js
    function tmp_visual() {
      if(g_img != undefined)
      {
        virtual_img.src = yarp.getImageSrc(g_img.compression_type,g_img.buffer);
        // video_stream_element.getContext('2d').drawImage(virtual_img,0,0,video_stream_element.width,video_stream_element.height);
        video_stream_element.getContext('2d').fillStyle = "#FF0000";
        video_stream_element.getContext('2d').fillRect(0, 0, video_stream_element.width, video_stream_element.height);
        var myImageData = video_stream_element.getContext('2d').getImageData(0, 0, video_stream_element.width, video_stream_element.height);
      }

      setTimeout(tmp_visual,33);
    }
```

**Comment by nunoguedelha** [comment]
## Check where in the transmission between the source file and final rendering the Color Space is modified

We define three test image files for checking color space changes between the source and destination images. Each test image is a rectangle filling the canvas, either red, green or blue. The test images are encoded in PPM P6 format, manually edited in P3 (ASCII image data) and then converted to P6 (binary image data) in Gimp (saved as "raw" format).

#### PPM format
http://paulbourke.net/dataformats/ppm/
https://web.cse.ohio-state.edu/~shen.94/681/Site/ppm_help.html
#### RGB color chart
https://www.rapidtables.com/web/color/RGB_Color.html

<details>
<summary>Example - Red rectangle</summary>

```
P3
# 8-bit ppm - RGB
10 10
255
255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0
255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0
255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0
255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0
255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0
255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0
255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0
255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0
255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0
255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0 255   0   0
```

</details>

**Comment by nunoguedelha** [comment]
## Check where in the transmission between the source file and final rendering the Color Space is modified [..Continued]

Let us recap the pipeline for rendering the rectangle image in the browser, either through the \<fakeFrameGrabber\> and \<YarpJS Example Server\>, either by loading directly the file in the browser (`virtual_img.src = "/testImage-xxxRectangle.jpg";`).

### Through the YarpJS stream video example

#### Config file **fakeFrameGrabber_testImage.ini**:
```
device fakeFrameGrabber
name /icubSim/camLeftEye
mode none
period 10
```
#### Run command:
```
yarpdev --file fakeFrameGrabber_testImage.ini --src testImage-xxxRectangle-P6.ppm
```
#### Transmission pipeline:

[testImage-xxxRectangle-P6.ppm] ⟶ \<fakeFrameGrabber\> ⟶ `/icubSim/camLeftEye` ⟶ `/yarpjs/img:i` ⟶ \<yarp.js example server\> ⟶ Yarp Socket ⟶ \<HTML page\>

- The JPEG data is generated in the \<fakeFrameGrabber\> device.
- In the \<yarp.js example server\> there are actually two steps: (1) reading the data through the Javascript bindings, (2) sending the data via the YarpJS Communicator.

#### Loading directly the file in the browser

[testImage-xxxRectangle.jpg] ⟶ \<HTML page\>

- The JPEG file was generated from the PPM original file using GIMP.


### Comparing the Image file data VS image data received through the Yarp socket 
We compared in NodeJS command line the image appearance and metadata from the file "testImage-xxxRectangle.jpg" against what we get from the image data received by the client.

**Comment by nunoguedelha** [comment]
## Check where in the transmission between the source file and final rendering the Color Space is modified [..On and On]

When reading the Yarp port `/icubSim/camLeftEye` data with the command
```
yarp read ... /icubSim/camLeftEye
```
We get the output
```
[mat] [rgb] (3 320 8 10 10) {200 0 100 200 0 100 ...)
```
This is the header `[mat] [rgb] (3 320 8 10 10)` and raw data (PPM color points) received on the port, read as a bottle. Note that the data sent from the **fakeFrameGrabber** is not compressed.

Focusing now on the analysis of the processing done in `YarpJS.node`...

**Comment by nunoguedelha** [comment]
## Check where in the transmission between the source file and final rendering the Color Space is modified [processing done in `YarpJS.node`]

Refer to #31 fro further details on the setup used for debugging `YarpJS.node` code on CLion.

### Received Data in YarpJS_BufferedPort_Image::_callback_onRead

We run the `YarpJS` target and break in
```js
YarpJS_BufferedPort_Image::_callback_onRead(std::vector<…> &) YarpJS_BufferedPort_Image.cpp:30
YarpJS_Callback::_internal_worker_end(uv_work_s *, int) YarpJS_Callback.h:123
```

https://github.com/robotology/yarp.js/blob/d2e4609c5115544c3837eb36fa38065aa06a343f/YarpJS/include/YarpJS_Callback.h#L109-L128

=> call to `tmp_this->prepareCallback`

https://github.com/robotology/yarp.js/blob/d2e4609c5115544c3837eb36fa38065aa06a343f/YarpJS/src/YarpJS_BufferedPort_Image.cpp#L20-L31

```c++
argv = {v8::Local<v8::Value> [1]} 
 [0] = {v8::Local<v8::Value>} 
  val_ = {v8::Value *} 0x1080080b8 
cons = {v8::Local<v8::Function>} 
tmpImgJS = {v8::Local<v8::Object>} 
 val_ = {v8::Object *} 0x105039638 
tmpImg = {YarpJS_Image *} 0x104f1cf80 
 YarpJS_Wrapper<yarp::sig::Image> = {YarpJS_Wrapper<yarp::sig::Image>} 
  YarpJS = {YarpJS} 
  yarpObj = {yarp::sig::Image *} 0x107307e90 
   imgWidth = {size_t} 10
   imgHeight = {size_t} 10
   imgPixelSize = {size_t} 3
   imgRowSize = {size_t} 32
   imgQuantum = {size_t} 8
   imgPixelCode = {int} 6449010
   topIsLow = {bool} true
   data = {char **} 0x1073070e0 
    *data = {char *} 0x107308528 "\xc8"
     **data = {char} -56 '\xc8'
```
In `YarpJS_BufferedPort_Image::_callback_onRead`, the data pointed by `tmpImg.﻿﻿YarpJS_Wrapper<yarp::sig::Image>.yarpObj->data` show:
<img width="837" alt="image" src="https://user-images.githubusercontent.com/6848872/162856662-5d2a6b31-48e4-4e18-8aa3-4f48bfde6a87.png">

The pixel data `0x c8 00 64` translates to decimal RGB colors `200 0 100` as expected. So the swap happens later on.

**Comment by nunoguedelha** [comment]
The analysis in https://github.com/robotology/yarp.js/issues/41#issuecomment-1096701942 narrowed down the bug location within the JPEG compression function called in https://github.com/robotology/yarp.js/blob/d2e4609c5115544c3837eb36fa38065aa06a343f/YarpJS/src/YarpJS_Image.cpp#L68.

This is enough to proceed with the performance analysis and improvement in https://github.com/ami-iit/yarp-openmct/issues/104 by trying a new transmission protocol (mjpeg), before actually implementing a fix here, since that issue has more priority and the compression used with the MJPEG protocol shall replace the JPEG compression implemented in `yarp.js`.

**Comment by nunoguedelha** [comment]
CC @traversaro

**Comment by nunoguedelha** [comment]
## Convert the images to OpenCV JPEG format

```c++
/** @brief Encodes an image into a memory buffer.

The function imencode compresses the image and stores it in the memory buffer that is resized to fit the
result. See cv::imwrite for the list of supported formats and flags description.

@param ext File extension that defines the output format.
@param img Image to be written.
@param buf Output buffer resized to fit the compressed image.
@param params Format-specific parameters. See cv::imwrite and cv::ImwriteFlags.
*/
CV_EXPORTS_W bool imencode( const String& ext, InputArray img,
                            CV_OUT std::vector<uchar>& buf,
                            const std::vector<int>& params = std::vector<int>());
```
```c++
/** @brief Saves an image to a specified file.

The function imwrite saves the image to the specified file. The image format is chosen based on the
filename extension (see cv::imread for the list of extensions). In general, only 8-bit
single-channel or 3-channel (with 'BGR' channel order) images
can be saved using this function, with these exceptions:

- 16-bit unsigned (CV_16U) images can be saved in the case of PNG, JPEG 2000, and TIFF formats
- 32-bit float (CV_32F) images can be saved in PFM, TIFF, OpenEXR, and Radiance HDR formats;
  3-channel (CV_32FC3) TIFF images will be saved using the LogLuv high dynamic range encoding
  (4 bytes per pixel)
- PNG images with an alpha channel can be saved using this function. To do this, create
8-bit (or 16-bit) 4-channel image BGRA, where the alpha channel goes last. Fully transparent pixels
should have alpha set to 0, fully opaque pixels should have alpha set to 255/65535 (see the code sample below).
- Multiple images (vector of Mat) can be saved in TIFF format (see the code sample below).

If the image format is not supported, the image will be converted to 8-bit unsigned (CV_8U) and saved that way.

If the format, depth or channel order is different, use
Mat::convertTo and cv::cvtColor to convert it before saving. Or, use the universal FileStorage I/O
functions to save the image to XML or YAML format.

The sample below shows how to create a BGRA image, how to set custom compression parameters and save it to a PNG file.
It also demonstrates how to save multiple images in a TIFF file:
@include snippets/imgcodecs_imwrite.cpp
@param filename Name of the file.
@param img (Mat or vector of Mat) Image or Images to be saved.
@param params Format-specific parameters encoded as pairs (paramId_1, paramValue_1, paramId_2, paramValue_2, ... .) see cv::ImwriteFlags
*/
CV_EXPORTS_W bool imwrite( const String& filename, InputArray img,
              const std::vector<int>& params = std::vector<int>());
```
```c++
enum ImwriteFlags {
       IMWRITE_JPEG_QUALITY        = 1,  //!< For JPEG, it can be a quality from 0 to 100 (the higher is the better). Default value is 95.
       IMWRITE_JPEG_PROGRESSIVE    = 2,  //!< Enable JPEG features, 0 or 1, default is False.
       IMWRITE_JPEG_OPTIMIZE       = 3,  //!< Enable JPEG features, 0 or 1, default is False.
       IMWRITE_JPEG_RST_INTERVAL   = 4,  //!< JPEG restart interval, 0 - 65535, default is 0 - no restart.
       IMWRITE_JPEG_LUMA_QUALITY   = 5,  //!< Separate luma quality level, 0 - 100, default is 0 - don't use.
       IMWRITE_JPEG_CHROMA_QUALITY = 6,  //!< Separate chroma quality level, 0 - 100, default is 0 - don't use.
       IMWRITE_PNG_COMPRESSION     = 16, //!< For PNG, it can be the compression level from 0 to 9. A higher value means a smaller size and longer compression time. If specified, strategy is changed to IMWRITE_PNG_STRATEGY_DEFAULT (Z_DEFAULT_STRATEGY). Default value is 1 (best speed setting).
       IMWRITE_PNG_STRATEGY        = 17, //!< One of cv::ImwritePNGFlags, default is IMWRITE_PNG_STRATEGY_RLE.
       IMWRITE_PNG_BILEVEL         = 18, //!< Binary level PNG, 0 or 1, default is 0.
       IMWRITE_PXM_BINARY          = 32, //!< For PPM, PGM, or PBM, it can be a binary format flag, 0 or 1. Default value is 1.
       IMWRITE_EXR_TYPE            = (3 << 4) + 0, /* 48 */ //!< override EXR storage type (FLOAT (FP32) is default)
       IMWRITE_EXR_COMPRESSION     = (3 << 4) + 1, /* 49 */ //!< override EXR compression type (ZIP_COMPRESSION = 3 is default)
       IMWRITE_WEBP_QUALITY        = 64, //!< For WEBP, it can be a quality from 1 to 100 (the higher is the better). By default (without any parameter) and for quality above 100 the lossless compression is used.
       IMWRITE_PAM_TUPLETYPE       = 128,//!< For PAM, sets the TUPLETYPE field to the corresponding string value that is defined for the format
       IMWRITE_TIFF_RESUNIT = 256,//!< For TIFF, use to specify which DPI resolution unit to set; see libtiff documentation for valid values
       IMWRITE_TIFF_XDPI = 257,//!< For TIFF, use to specify the X direction DPI
       IMWRITE_TIFF_YDPI = 258, //!< For TIFF, use to specify the Y direction DPI
       IMWRITE_TIFF_COMPRESSION = 259, //!< For TIFF, use to specify the image compression scheme. See libtiff for integer constants corresponding to compression formats. Note, for images whose depth is CV_32F, only libtiff's SGILOG compression scheme is used. For other supported depths, the compression scheme can be specified by this flag; LZW compression is the default.
       IMWRITE_JPEG2000_COMPRESSION_X1000 = 272 //!< For JPEG2000, use to specify the target compression rate (multiplied by 1000). The value can be from 0 to 1000. Default is 1000.
     };
```

The current parameters `p` are
https://github.com/robotology/yarp.js/blob/d2e4609c5115544c3837eb36fa38065aa06a343f/YarpJS/src/YarpJS_Image.cpp#L34-L35

No additional parameter here is meant to convert the channel order.

> If the format, depth or channel order is different, use
> Mat::convertTo and cv::cvtColor to convert it before saving. ...
> 

### cvtColor(...)

> /** @brief Converts an image from one color space to another.
> 
> The function converts an input image from one color space to another. In case of a transformation
> to-from RGB color space, the order of the channels should be specified explicitly (RGB or BGR). Note
> that the default color format in OpenCV is often referred to as RGB but it is actually BGR (the
> bytes are reversed). So the first byte in a standard (24-bit) color image will be an 8-bit Blue
> component, the second byte will be Green, and the third byte will be Red. The fourth, fifth, and
> sixth bytes would then be the second pixel (Blue, then Green, then Red), and so on.
> 
> The conventional ranges for R, G, and B channel values are:
> -   0 to 255 for CV_8U images
> -   0 to 65535 for CV_16U images
> -   0 to 1 for CV_32F images
> 
> In case of linear transformations, the range does not matter. But in case of a non-linear
> transformation, an input RGB image should be normalized to the proper value range to get the correct
> results, for example, for RGB \f$\rightarrow\f$ L\*u\*v\* transformation. For example, if you have a
> 32-bit floating-point image directly converted from an 8-bit image without any scaling, then it will
> have the 0..255 value range instead of 0..1 assumed by the function. So, before calling #cvtColor ,
> you need first to scale the image down:
> @code
>     img *= 1./255;
>     cvtColor(img, img, COLOR_BGR2Luv);
> @endcode
> If you use #cvtColor with 8-bit images, the conversion will have some information lost. For many
> applications, this will not be noticeable but it is recommended to use 32-bit images in applications
> that need the full range of colors or that convert an image before an operation and then convert
> back.
> 
> If conversion adds the alpha channel, its value will set to the maximum of corresponding channel
> range: 255 for CV_8U, 65535 for CV_16U, 1 for CV_32F.
> 
> @param src input image: 8-bit unsigned, 16-bit unsigned ( CV_16UC... ), or single-precision
> floating-point.
> @param dst output image of the same size and depth as src.
> @param code color space conversion code (see #ColorConversionCodes).
> @param dstCn number of channels in the destination image; if the parameter is 0, the number of the
> channels is derived automatically from src and code.
> 
> @see @ref imgproc_color_conversions
>  */
> CV_EXPORTS_W void cvtColor( InputArray src, OutputArray dst, int code, int dstCn = 0 );
> 

Tried a fix replacing
https://github.com/robotology/yarp.js/blob/d2e4609c5115544c3837eb36fa38065aa06a343f/YarpJS/src/YarpJS_Image.cpp#L39

by...
```c++
    cv::Mat      internalImageBGR;
    cv::cvtColor(internalImage,internalImageBGR,cv::COLOR_RGB2BGR,0);
    cv::imencode(encodeString,internalImageBGR, internalBuffer, p);
```

**Comment by nunoguedelha** [comment]
As per Stefano's suggestion, check if we can use directly the automatic conversion from Yarp.

**Comment by nunoguedelha** [comment]
> As per @S-Dafarra's suggestion, check if we can use directly the automatic conversion from Yarp.

Shall be handled in #43 .

**Comment by nunoguedelha** [comment]
switched RGB colors fixed by #40 .
